src/API/Recruit.Api/Commands/CreateTraineeshipVacancyCommandHandler.cs

```diff
@@ -0,0 +1,231 @@
+
+
+            var result = _recruitVacancyClient.Validate(request.Vacancy, VacancyRuleSet.All);
+
+            if (result.HasErrors)
+            {
+                return new CreateTraineeshipVacancyCommandResponse
+                {
+                    ResultCode = ResponseCode.InvalidRequest,
+                    ValidationErrors = result.Errors.Select(error => new DetailedValidationError
+                    {
+                        Field = error.PropertyName,
+                        Message = error.ErrorMessage
+                    }).Cast<object>().ToList()
+                };
+            }
+
+            if (request.ValidateOnly)
+            {
+                return new CreateTraineeshipVacancyCommandResponse
+                {
+                    ResultCode = ResponseCode.Created,
+                    Data = 1000000001
+                };
+            }
+
+            try
+            {
+                await CreateVacancy(request, trainingProvider);
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Error creating vacancy");
+                return new CreateTraineeshipVacancyCommandResponse
+                {
+                    ResultCode = ResponseCode.InvalidRequest,
+                    ValidationErrors = new List<DetailedValidationError>{new DetailedValidationError
+                    {
+                        Field = nameof(request.Vacancy.Id), Message = "Unable to create Vacancy. Vacancy already submitted"
+                    }}.Cast<object>().ToList()
+                };
+            }
+
+            var requiresEmployerApproval = await CheckEmployerApprovalNeeded(request);
+
+            var newVacancy = await MapDraftVacancyValues(request, request.Vacancy, requiresEmployerApproval);
+
+            await _vacancyRepository.UpdateAsync(newVacancy);
+
+            await PublishVacancyEvent(requiresEmployerApproval, newVacancy);
+
+            return new CreateTraineeshipVacancyCommandResponse
+            {
+                ResultCode = ResponseCode.Created,
+                Data = newVacancy.VacancyReference.Value
+            };
+        }
+
+        private async Task CreateVacancy(CreateTraineeshipVacancyCommand request, TrainingProvider trainingProvider)
+        {
+            if (!string.IsNullOrEmpty(request.VacancyUserDetails.Email))
+            {
+                request.VacancyUserDetails.Ukprn = null;
+                await _employerVacancyClient.CreateEmployerApiVacancy(request.Vacancy.Id, request.Vacancy.Title,
+                    request.Vacancy.EmployerAccountId,
+                    request.VacancyUserDetails, trainingProvider, request.Vacancy.ProgrammeId);
+            }
+            else
+            {
+                await _providerVacancyClient.CreateProviderApiVacancy(request.Vacancy.Id, request.Vacancy.Title,
+                    request.Vacancy.EmployerAccountId,
+                    request.VacancyUserDetails);
+            }
```

Comment:
This can be simplified - only providers are allowed to create Traineeship vacancies